### ISSUE TITLE
[check] `RandomLiteralEncoding`

### ISSUE THREAD

**Issue by Pankraz76** [google/error-prone#5378]
considering:

- https://github.com/gradle/gradle/pull/35871

# [check] `RandomLiteralEncoding`

Many codebases still rely on **hard-coded encoding strings** such as `"utf-8"`, `"UTF8"`, `"Utf-8"`, or `"utf8"`.  
Although these may *appear* harmless, they are in practice **error-prone**, **inconsistent**, and entirely **obsolete** when robust, well-defined constants already exist.

Manually typing encodings opens the door to a steady stream of small but disruptive mistakes:

- `"UTF8"` — missing the required dash  
- `"Utf-8"` — unpredictable capitalization  
- `"utf8"` — accepted by some tools, rejected by others  
- `"utf-8"` vs `"UTF-8"` — semantically identical, but creates noisy diffs

These inconsistencies add no value, introduce subtle bugs, and increase maintenance burden.  
And crucially: **all of this is avoidable.**

---

## Purpose

The `RandomLiteralEncoding` check identifies **literal string encodings** in source code and enforces the use of **standard charset constants** such as `StandardCharsets.UTF_8`.  
Using constants eliminates typos, prevents platform-dependent fallbacks, reduces diff noise, and ensures encoding usage is clear and intentional.

---

## What the check flags

The check reports **string literals** that match known encoding patterns when passed into APIs expecting charset information, for example:

```java
"utf-8"
"UTF-8"
"UtF-8"
"uTF-8"
"Utf-8"
"us-ascii"
"latin1"
"ISO_8859-1"
and so on....
```

**Comment by tbroyer** [comment]
Doesn't https://errorprone.info/bugpattern/StringCharset already do that?

**edit:** oh, you'd like to flag any use of a string literal that looks like a charset name, even (particularly) in places where no overload accepts a Charset (so places where static analysis cannot determine that these string literals actually are charset names)

**Comment by Pankraz76** [comment]
yes im sorry was also thinking about having stumbled upon this already. But as we have activated it already somehow, im confused about it.

why is one active but the other one not, or we we have so many blindspots rather in gradle config or code?

- https://github.com/gradle/gradle/pull/35863
- https://github.com/gradle/gradle/pull/34987
- https://github.com/search?q=repo%3Agradle%2Fgradle+DefaultCharset+&type=code
- https://github.com/search?q=repo%3Agradle%2Fgradle%20StringCharset%20&type=code
